src/charm.py

```diff
@@ -63,6 +82,95 @@ def _on_website_relation_joined(self, event):
             'port': str(port)
         })
 
+    def _setup_promtail(self):
+        # Download promtail binary
+        check_call(["curl", "-LO", "https://github.com/grafana/loki/releases/download/v2.6.1/promtail-linux-amd64.zip"])
+        check_call(["busybox", "unzip", "promtail-linux-amd64.zip"])
+        check_call(["sudo", "mv", "promtail-linux-amd64", "/usr/local/bin/promtail"])
+        check_call(["sudo", "mkdir", "-p", "/etc/promtail"])
+
+        # Create promtail service
+        svc_file = open("/etc/systemd/system/promtail.service", "w")
+        svc_file.write('''[Unit]
+Description=Promtail service
+After=network.target
+
+[Service]
+Type=simple
+User=root
+ExecStart=/usr/local/bin/promtail -config.file /etc/promtail/config.yaml
+
+[Install]
+WantedBy=multi-user.target
+''')
+        svc_file.close()
+        check_call(["sudo", "systemctl", "daemon-reload"])
+
+        # Start promtail service
+        self._restart_promtail()
+
+    def _restart_promtail(self, _ = None):
+        '''Reconfigures and restarts the promtail service when the logging
+        relation is changed.
+        Assumes the promtail service has already been set up by _setup_promtail
+        '''
+
+        # Find client push urls
+        endpoints = []
+        for relation in self.model.relations[LOGGING_RELATION_NAME]:
+            for unit in relation.units:
+                ip = relation.data[unit].get("private-address")
+                if ip:
+                    endpoints.append({'url': f'http://{ip}:3100/loki/api/v1/push'})
+
+        if len(endpoints) == 0:
+            # There are no targets, so we can't run Promtail
+            # Make sure the Promtail service is stopped
+            check_call(["sudo", "systemctl", "stop", "promtail.service"])
+        else:
+            # Create Promtail config object
+            config = self._promtail_base_config()
+            config['clients'] = endpoints
+
+            # Write config file
+            cfg_file = open("/etc/promtail/config.yaml", "w")
+            yaml.dump(config, cfg_file)
+            cfg_file.close()
+
+            # Start/restart Promtail
+            check_call(["sudo", "systemctl", "restart", "promtail.service"])
+
+
+    def _promtail_base_config(self):
+        '''Returns base configuration for Promtail (to be serialised to yaml)'''
```

Comment:
some of these should probably be charm config (with sensible defaults)
especially the scrape config. logsink is a good start i think. do we really need everything in /var/log? syslog maybe to start with?

Reply:
Yeah this is definitely something to think more about. I'm not sure which logs people would want to see. We can start with logsink and syslog.

Reply:
So the `server` option just specifies ports for a "Promtail server" which can be used for simple debugging. We actually don't even need to specify these. The default ports are
```
http=[::]:80 grpc=[::]:9095
```